packages/server/routes/api/images.js

```diff
@@ -0,0 +1,243 @@
+const express = require("express");
+const router = express.Router();
+const { validate } = require("uuid");
+
+const AuthController = require("../../controllers/auth");
+const ImagesController = require("../../controllers/images");
+
+const VALID_IMAGE_TYPES = [
+  "user_icon",
+  "guild_icon",
+  "guild_banner",
+  "event_banner",
+];
+
+router.put(
+  "/:type/:UUID",
+  AuthController.authenticate,
+  async (request, response) => {
+    const imageType = request.params.type;
+    const imageUUID = request.params.UUID;
+    const imageData = request.body.imageData;
+
+    if (!VALID_IMAGE_TYPES.includes(imageType)) {
+      response.status(400).json({
+        status: "fail",
+        data: {
+          imageType: `Image type '${imageType}' is invalid.`,
+        },
+      });
+      return;
+    }
+    if (!validate(imageUUID)) {
+      response.status(400).json({
+        status: "fail",
+        data: {
+          imageUUID: `Image UUID '${imageUUID}' is invalid.`,
+        },
+      });
+      return;
+    }
+    if (!imageData) {
+      response.status(400).json({
+        status: "fail",
+        data: {
+          imageData: "Request body missing value 'imageData'.",
+        },
+      });
+      return;
+    }
+    if (!imageData.startsWith("/9j/")) {
+      response.status(400).json({
+        status: "fail",
+        data: {
+          imageData: "Invalid Base64-encoded JPEG.",
+        },
+      });
+      return;
+    }
+    if (!(await ImagesController.exists(imageType, imageUUID))) {
+      response.status(400).json({
+        status: "fail",
+        data: {
+          imageUUID: `A ${imageType} with UUID ${imageUUID} does not exist.`,
+        },
+      });
+      return;
+    }
+
+    try {
+      const url = await ImagesController.upload(
+        imageType,
+        imageUUID,
+        imageData
+      );
+      response.status(200).json({
+        status: "success",
+        data: {
+          imageURL: url,
+        },
+      });
+    } catch (err) {
+      response.status(500).json({
+        status: "error",
+        message: err.message,
+      });
+    }
+  }
+);
+
+router.get(
+  "/:type/:UUID",
+  AuthController.authenticate,
+  async (request, response) => {
+    const imageType = request.params.type;
+    const imageUUID = request.params.UUID;
+
+    if (!VALID_IMAGE_TYPES.includes(imageType)) {
+      response.status(400).json({
+        status: "fail",
+        data: {
+          imageType: `Image type '${imageType}' is invalid.`,
+        },
+      });
+      return;
+    }
+    if (!validate(imageUUID)) {
+      response.status(400).json({
+        status: "fail",
+        data: {
+          imageUUID: `Image UUID '${imageUUID}' is invalid.`,
+        },
+      });
+      return;
+    }
+
+    try {
+      const s3response = await ImagesController.retrieve(imageType, imageUUID);
+      response.status(200);
+      response.send(s3response);
```

Comment:
I tested fetching a nonexistent image in the S3 storage and it gave me a dead link and was still considered a successful response. The API route should throw an error if the frontend tries to fetch a nonexistent image with the GET route. You'll have to change `ImagesController.retrieve` to throw an error if there is no image in S3 that exists with the given key.